### PR TITLE
New hlsl mpl

### PR DESCRIPTION
**Pull request by Przemog1**

### CODE REVIEW

File: include/nbl/builtin/hlsl/mpl.hlsl [hunk 1]
```diff
@@ -24,21 +24,24 @@ struct clz_masks
...
+//     static const uint64_t LO_MASK = 0;
+// };
+
 template<uint64_t N, uint16_t bits_log2>
 struct clz
 {
     static const bool CHOOSE_HIGH = N&(clz_masks<bits_log2>::LO_MASK<<clz_masks<bits_log2>::SHIFT);
     static const uint64_t NEXT_N = (CHOOSE_HIGH ? (N>>clz_masks<bits_log2>::SHIFT):N)&clz_masks<bits_log2>::LO_MASK;
-    static const uint16_t value   = type_traits::conditional<bits_log2,clz<NEXT_N,bits_log2-1>,type_traits::integral_constant<uint16_t,0> >::type::value + (CHOOSE_HIGH ? 0ull:clz_masks<bits_log2>::SHIFT);
+    static const uint16_t value = type_traits::conditional<bits_log2,clz<NEXT_N,bits_log2-1>,type_traits::integral_constant<uint16_t,0> >::type::value + (CHOOSE_HIGH ? 0ull:clz_masks<bits_log2>::SHIFT);
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/mpl.hlsl:44]: can you make that `conditional_value` so we can shorten expressions like this?

Also I'd still have a `clz_masks` and have the conditional live there.

File: include/nbl/builtin/hlsl/mpl.hlsl [hunk 2]
```diff
@@ -41,7 +41,8 @@ struct countl_zero
 {
     NBL_CONSTEXPR_STATIC_INLINE bool CHOOSE_HIGH = N&(countl_zero_masks<bits_log2>::LO_MASK<<countl_zero_masks<bits_log2>::SHIFT);
     NBL_CONSTEXPR_STATIC_INLINE uint64_t NEXT_N = (CHOOSE_HIGH ? (N>>countl_zero_masks<bits_log2>::SHIFT):N)&countl_zero_masks<bits_log2>::LO_MASK;
-    NBL_CONSTEXPR_STATIC_INLINE uint16_t value   = type_traits::conditional<bits_log2,countl_zero<NEXT_N,bits_log2-1>,type_traits::integral_constant<uint16_t,0> >::type::value + (CHOOSE_HIGH ? 0ull:countl_zero_masks<bits_log2>::SHIFT);
+    //NBL_CONSTEXPR_STATIC_INLINE uint16_t value = type_traits::conditional_value<bits_log2,uint16_t,countl_zero<NEXT_N,bits_log2-1>::value,type_traits::integral_constant<uint16_t,0>::value>::value + (CHOOSE_HIGH ? 0ull:countl_zero_masks<bits_log2>::SHIFT);
+    NBL_CONSTEXPR_STATIC_INLINE uint16_t value = type_traits::conditional<bits_log2,countl_zero<NEXT_N,bits_log2-1>,type_traits::integral_constant<uint16_t,0> >::type::value + (CHOOSE_HIGH ? 0ull:countl_zero_masks<bits_log2>::SHIFT);
```
Comment (devshgraphicsprogramming) [inline include/nbl/builtin/hlsl/mpl.hlsl:45]: well use the `conditional_value` somewhere :D

Comment (Przemog1) [reply include/nbl/builtin/hlsl/mpl.hlsl:45]: if i uncomment this line and use `conditional_value` then it says that `NEXT_N` is not a constant expression which doesn't make sense to me..

Comment (devshgraphicsprogramming) [reply include/nbl/builtin/hlsl/mpl.hlsl:45]: you mean when the line becomes
```cpp
NBL_CONSTEXPR_STATIC_INLINE uint16_t value = type_traits::conditional_value<bits_log2,uint16_t,countl_zero<NEXT_N,bits_log2-1>::value,0>::value + (CHOOSE_HIGH ? 0ull:countl_zero_masks<bits_log2>::SHIFT);
```
?

Comment (Przemog1) [reply include/nbl/builtin/hlsl/mpl.hlsl:45]: yes

**Pull request merged**